### ISSUE TITLE
Pango 1.50.12+ cannot load non-system fonts for Wesnoth 1.16 on Windows

### ISSUE THREAD

**Issue by Wedge009** [wesnoth/wesnoth#7543]
### Game and System Information

- **Version:** 1.16.9
- **Downloaded from:** Steam, but I expect other Windows releases to exhibit the same problem.
- **Build info:** Official release build
- **OS:** Windows 7 and 10

### Description of the bug

Pango reports being unable to load Lato font as per logs:
```
(process:11320): Pango-WARNING **: 09:21:34.812: couldn't load font "Lato Not-Rotated 20", falling back to "Sans Not-Rotated 20", expect ugly output.

(process:11320): Pango-WARNING **: 09:21:34.824: couldn't load font "Lato 14", falling back to "Sans 14", expect ugly output.

(process:11320): Pango-WARNING **: 09:21:34.840: couldn't load font "Lato Not-Rotated 17", falling back to "Sans Not-Rotated 17", expect ugly output.
Checking lua scripts... ok

(process:11320): Pango-WARNING **: 09:21:35.269: couldn't load font "Lato Italic Not-Rotated 17", falling back to "Sans Italic Not-Rotated 17", expect ugly output.

(process:11320): Pango-WARNING **: 09:21:35.275: couldn't load font "Lato Bold Not-Rotated 17", falling back to "Sans Bold Not-Rotated 17", expect ugly output.

(process:11320): Pango-WARNING **: 09:21:35.307: couldn't load font "Lato Not-Rotated 16", falling back to "Sans Not-Rotated 16", expect ugly output.

(process:11320): Pango-WARNING **: 09:21:35.311: couldn't load font "Lato Not-Rotated 15", falling back to "Sans Not-Rotated 15", expect ugly output.

(process:11320): Pango-WARNING **: 09:21:35.313: couldn't load font "Lato Not-Rotated 13", falling back to "Sans Not-Rotated 13", expect ugly output.

(process:11320): Pango-WARNING **: 09:21:35.317: couldn't load font "Lato Not-Rotated 14", falling back to "Sans Not-Rotated 14", expect ugly output.
```

So far, I only see this issue on 1.16 releases for Windows. 1.17 doesn't appear affected, neither are Linux releases.

1.16.9 on Windows 7:
![169](https://user-images.githubusercontent.com/13679322/232349974-097db2fb-c077-4fa5-a0db-5261b7f56a2c.jpg)

### Steps to reproduce the behavior

Run Wesnoth 1.16.9 on Windows.

### Expected behavior

1.16.8 on Windows 7:
![168](https://user-images.githubusercontent.com/13679322/232350131-32e7e837-38a2-48e9-a312-59b7982cadf6.jpg)

1.17.15 on Windows 7:
![1715](https://user-images.githubusercontent.com/13679322/232350185-7e4e9a4d-8d38-4f9d-8836-1910e730a3cd.jpg)

1.16.9 on Linux:
![169-linux](https://user-images.githubusercontent.com/13679322/232350440-b699d374-b14f-4779-914a-a279666cdba9.jpg)

### Additional context

Don't know if it's relevant but Windows release builds includes pango 1.50.11 for Wesnoth 1.16.8 and pango 1.50.14 for both 1.16.9 and 1.17.15.

I cannot bisect this issue due to #7217.

**Comment by Pentarctagon** [comment]
Are there Lato fonts present in wesnoth's `fonts/` directory?

**Comment by Wedge009** [comment]
Yes, I did check they were all there - seems to be unchanged from 1.16.8.

**Comment by Pentarctagon** [comment]
Darn, I was hoping it was that simple...

@Vultraz or maybe @knyghtmare would you be able to try bisecting this?

**Comment by Wedge009** [comment]
> Darn, I was hoping it was that simple...

Was one of the first things I checked, also hoping to find a straightforward cause...

**Comment by Wedge009** [comment]
Labelled as possible upstream issue as it seems to be pango-related. I dropped pango 1.50.11 DLLs (`libpango*.dll` - four files) from 1.16.8 release into 1.16.9 release and the Lato font seems to load fine.

Although this doesn't explain why Wesnoth 1.17.15 doesn't have the issue despite using the same pango version (I believe) as Wesnoth 1.16.9.

**Comment by knyghtmare** [comment]
>  would you be able to try bisecting this?

What am I supposed to look for?

Should I use 1.16.8 as a starting point and the latest 1.16.x commit as the end point?

**Comment by knyghtmare** [comment]
![image](https://user-images.githubusercontent.com/5283677/233081355-48eaa696-e6cd-4eaf-a9e6-1807d8643b2b.png)

I cannot seem to switch to 1.16.9+dev branch on Github Desktop. This was working fine...What is this?

**Comment by Pentarctagon** [comment]
@knyghtmare I guess it doesn't like the lua modules directory not existing for 1.16 for some reason. I don't know why though - a simple `git checkout 1.16` on the command line works fine.

**Comment by Zireael07** [comment]
Github desktop client has always been buggy for me. Try another free GUI git client - GitKraken and SourceTree are both good in my experience

**Comment by gfgtdf** [comment]
So if i understood the comment in https://github.com/wesnoth/wesnoth/pull/7544#issuecomment-1514846027 correctly we fixed the 1.16.9 windows steam/download release by using oder dlls version? And this is 'only' about self builds and 1.16.10+ now?

**Comment by Pentarctagon** [comment]
Correct.

**Comment by Wedge009** [comment]
It appears that with the way 1.16 uses Pango - without fontconfig that's currently in master - something changed in pango between 1.50.11 and 1.50.14 such that on Windows platforms it will only load a font if it's available in the Windows system pool of fonts. I confirmed this by installing the Lato fonts on my system, and re-confirmed the problem after uninstalling the fonts.

I found this from browsing various reports of the same issue for projects that depend on Pango, not Pango itself. Looking at [Pango's change log](https://gitlab.gnome.org/GNOME/pango/-/blob/main/NEWS), there doesn't appear to be anything directly associated with this change in behaviour.

**Comment by Pentarctagon** [comment]
So this is potentially an upstream issue, in that case?

**Comment by Wedge009** [comment]
Perhaps a change in how downstream users are expected to use it.

**Comment by Pentarctagon** [comment]
Did you find anything indicating what that new expected usage is? If not, it would probably be worth it to open an issue on the pango gitlab asking if it's intentional and if it is then what's the new expected way to handle this.

**Comment by Wedge009** [comment]
Actually, I'm not even sure that it's necessarily a new problem. I don't really know how Pango, Cairo, fontconfig, and Wesnoth are all supposed to interact. But reading through various issues like the following seems to suggest that it's an expectation to either use fontconfig to set-up the fonts or have the fonts installed and available system-wide:

https://github.com/Automattic/node-canvas/issues/1608
https://github.com/lovell/sharp/issues/1162#issuecomment-373951732
https://github.com/shoes/shoes3/issues/48

**Comment by Pentarctagon** [comment]
Reading through those, it sounds like there are a couple options to try and get some additional information:
* Try setting the `FC_DEBUG` environment variable (apparently `FC_DEBUG=8191` is the most verbose?)
* Add a glib log handler as shown in https://stackoverflow.com/a/26618242

**Comment by gfgtdf** [comment]
> But reading through various issues...

All of these issues are much older then this bug the pango version that changed this is from 2023 or late 2022, and even if our method is not listed there it doesn'tean that it's not supported.

I really think the only way to get clarity on this issue is to file an upstream bug report, ideally with a simple example code.

**Comment by Wedge009** [comment]
> All of these issues are much older then (sic) this bug the pango version that changed this is from 2023 or late 2022, and even if our method is not listed there it doesn'tean (sic) that it's not supported.

I noted that as well. As a quick test, I tried libpango 1.50.14 (my distribution normally uses 1.50.6) on Linux and confirmed I don't already have Lato installed as a system font - no apparent issues on either 1.17.18+dev nor 1.16.9 release nor 1.16.10+dev. Just confirming: Windows is the only platform where Wesnoth doesn't use fontconfig (and only for 1.16)?

**Comment by Pentarctagon** [comment]
Correct.

**Comment by Wedge009** [comment]
After some more testing with vcpkg's builds of Pango and peeking into the Pango source code and git blame records, I'm fairly confident this is caused by https://gitlab.gnome.org/GNOME/pango/-/commit/5bc16f7896ba3eb5b37ffdb03f28c646aad5889e. There's an issue already open for this: https://gitlab.gnome.org/GNOME/pango/-/issues/720

**Comment by Wedge009** [comment]
@loonycyborg Do you need to be aware of this? Perhaps we can hold to libpango 1.50.11 for the time being?

**Comment by Wedge009** [comment]
Advice from Pango developers seems to be to stay on libpango 1.50.11 until they replace the now-missing GDI functionality in a 'PangoWin32 API'. DirectWrite - which seems to ignore non-system installed fonts - replaced GDI usage in libpango 1.50.12.

https://gitlab.gnome.org/GNOME/pango/-/issues/720#note_1794746

**Comment by gfgtdf** [comment]
Nice, thx for finding this.

**Comment by loonycyborg** [comment]
I guess we would have to make a separate docker image for msys2+windows build then, with downgraded pango.

**Comment by Konrad22** [comment]
Hey, it seems the bug made it into the 1.16.10 release. :/

**Comment by Pentarctagon** [comment]
Should be fixed for 1.16.10 now.

**Comment by fanc999-1** [comment]
Hi,

This issue is raised as Pango/Cairo is transitioning to DirectWrite, so `AddFontResourceEx()` with `FR_PRIVATE` will be affected when using PangoWin32.

Related upstream items:
https://gitlab.gnome.org/GNOME/pango/-/issues/720
https://gitlab.gnome.org/GNOME/pango/-/imerge_requests/700
https://gitlab.gnome.org/GNOME/pango/-/imerge_requests/707

Current upstream Pango has a new API that works for Windows 10 and later that added support using an API adding non-installed fonts into the `PangoWin32FontMap`, and the version for Windows 7/8.x is under review/progress.

With blessings, and cheers!

**Comment by Pentarctagon** [comment]
Is there documentation for this?

**Comment by Wedge009** [comment]
I suppose this can be closed once 1.18 releases

**Comment by Wedge009** [comment]
Not an issue with 1.18.x release.
